extensions/gitpod-remote/src/extension.ts

```diff
@@ -0,0 +1,225 @@
+	if (vscode.workspace.workspaceFolders) {
+		return false;
+	}
+	const workspaceUri = vscode.Uri.file(context.info.getWorkspaceLocationFile() || context.info.getWorkspaceLocationFolder());
+	vscode.commands.executeCommand('vscode.openFolder', workspaceUri, { forceReuseWindow: true });
+	return true;
+}
+
+export async function installInitialExtensions(context: GitpodExtensionContext): Promise<void> {
+	context.logger.info('installing initial extensions...');
+	const extensions: (vscode.Uri | string)[] = [];
+	try {
+		const workspaceContextUri = vscode.Uri.parse(context.info.getWorkspaceContextUrl());
+		extensions.push('redhat.vscode-yaml');
+		if (/github\.com/i.test(workspaceContextUri.authority)) {
+			extensions.push('github.vscode-pull-request-github');
+		}
+
+		let config: { vscode?: { extensions?: string[] } } | undefined;
+		try {
+			const configUri = vscode.Uri.file(path.join(context.info.getCheckoutLocation(), '.gitpod.yml'));
+			const buffer = await vscode.workspace.fs.readFile(configUri);
+			const content = new util.TextDecoder('utf8').decode(buffer);
+			const model = new context.config.GitpodPluginModel(content);
+			config = model.document.toJSON();
+		} catch { }
+		if (config?.vscode?.extensions) {
+			const extensionIdRegex = /^([^.]+\.[^@]+)(@(\d+\.\d+\.\d+(-.*)?))?$/;
+			for (const extension of config.vscode.extensions) {
+				let link: vscode.Uri | undefined;
+				try {
+					link = vscode.Uri.parse(extension.trim(), true);
+					if (link.scheme !== 'http' && link.scheme !== 'https') {
+						link = undefined;
+					}
+				} catch { }
+				if (link) {
+					extensions.push(link);
+				} else {
+					const normalizedExtension = extension.toLocaleLowerCase();
+					if (extensionIdRegex.exec(normalizedExtension)) {
+						extensions.push(normalizedExtension);
+					}
+				}
+			}
+		}
+	} catch (e) {
+		context.logger.error('failed to detect workspace context dependent extensions:', e);
+		console.error('failed to detect workspace context dependent extensions:', e);
+	}
+	context.logger.info('initial extensions:', extensions);
+	if (extensions.length) {
+		let cause;
+		try {
+			const { stderr } = await util.promisify(cp.exec)('code ' + extensions.map(extension => '--install-extension ' + extension).join(' '));
+			cause = stderr;
+		} catch (e) {
+			cause = e;
+		}
+		if (cause) {
+			context.logger.error('failed to install initial extensions:', cause);
+			console.error('failed to install initial extensions: ', cause);
+		}
+	}
+	context.logger.info('initial extensions installed');
+}
+
+export function registerHearbeat(context: GitpodExtensionContext): void {
+	let lastActivity = 0;
+	const updateLastActivitiy = () => {
+		lastActivity = new Date().getTime();
+	};
+	const sendHeartBeat = async (wasClosed?: true) => {
+		const suffix = wasClosed ? 'was closed heartbeat' : 'heartbeat';
+		try {
+			// context.logger.info('sending ' + suffix);
```

Comment:
it can be trace level?

BTW how do I enable tracing in gitpod desktop extension?